src/java.base/share/classes/jdk/crac/impl/AbstractContextImpl.java

```diff
@@ -22,86 +22,178 @@
 
 package jdk.crac.impl;
 
-import jdk.crac.CheckpointException;
-import jdk.crac.Context;
-import jdk.crac.Resource;
-import jdk.crac.RestoreException;
-import sun.security.action.GetBooleanAction;
+import jdk.crac.*;
 
 import java.util.*;
-import java.util.stream.Collectors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
 
-public abstract class AbstractContextImpl<R extends Resource, P> extends Context<R> {
-
-    private static class FlagsHolder {
-        public static final boolean DEBUG =
-            GetBooleanAction.privilegedGetProperty("jdk.crac.debug");
-    }
-
-    private WeakHashMap<R, P> checkpointQ = new WeakHashMap<>();
+public abstract class AbstractContextImpl<R extends Resource> extends Context<R> {
+    private WeakHashMap<R, Long> resources = new WeakHashMap<>();
+    // Queue content is temporary, so we won't mind that it's not a weak reference
+    private Queue<Map.Entry<R, Long>> resourceQueue = new LinkedList<>();
     private List<R> restoreQ = null;
-    private Comparator<Map.Entry<R, P>> comparator;
+    private volatile long currentPriority = -1;
+    // We use two locks: checkpointLock is required for both running the beforeCheckpoint
+    // and registering a new resource, while restoreLock is required for running afterRestore
+    // and beforeCheckpoint (to achieve exclusivity of before and after). It is fine
+    // to acquire checkpointLock and register a new resource during afterRestore.
+    private final ReentrantLock checkpointLock = new ReentrantLock();
+    private final ReentrantLock restoreLock = new ReentrantLock();
 
-    protected AbstractContextImpl(Comparator<Map.Entry<R, P>> comparator) {
-        this.comparator = comparator;
+    protected void register(R resource, long priority) {
+        assert priority >= 0;
+        boolean locked = false;
+        try {
+            // We don't want to deadlock if the registration happens from another thread
+            while (!checkpointLock.tryLock(10, TimeUnit.MILLISECONDS)) {
+                throwIfCheckpointInProgress(priority);
+            }
+            locked = true;
+            // This is important for the case of recursive registration
+            throwIfCheckpointInProgress(priority);
```

Comment:
What is expected handling of this exception? And in the current form the exception is not checked, so in most cases that exception won't be expected. Having that in most cases resources are registered during class or object initialization, those entities will be caught in partially constructed states, likely leaving the whole system stuck.

Would it not better to allow registration, but throw CheckpointException at the end of Context.beforeCheckpoint? That is pretty legit exception that means the checkpoint can be attempted again, and another attempt can be pretty well successful, if all Resources do not throw and no new Resources are registered.

Reply:
So, making `register` throw a checked exception wouldn't be a good solution; most of the time there wouldn't be a better handling than to log && rethrow.

What you propose - having a form of exception stack in the context and checking it anything was added after `beforeCheckpoint` could work, though this is definitely not a pattern that would be natural in Java - it smells of JNI developer :)

I think that a situation when the ISE is thrown is a programming error, rather than reacting on the state of 'outer world' (as in case of e.g. I/O...) and as such this can stay as unchecked exception, possibly leading to inconsistent state of the program.

Reply:
Generally speaking, your proposal to try performing the checkpoint until it succeeds is more permissive, as users can back-off from a bad/fragile design through retrying, rather than being forced to correct the dependencies between components. In the past it seemed to me that you were more on the "good design required" rather than "practical" side of things - please don't take this as any form of "ad hominem" argument, I would just like to understand why you opt for such solution here. In my view a timeout/IO error are a totally different category of error.

Reply:
The CheckpointException is a CRaC-specific, deliberately thrown exception to communicate the problem with the state to the requester of the checkpoint. The exception may suggest rewriting the code, or another attempt. We don't care a lot about the preceise meaning, we just need to deliver the exception. Although lesser number of exceptions is better of course.

Reply:
I realized that you actually asked for a different behaviour than what I thought initially, something like the `ConcurrentModificationException` while iterating a collection. That is not that bad, and we can add debug option to track where this happened.

Reply:
Thinking about this again, the current code does not prevent that from happening either, had JDK context registered something new in the global context this would be silently ignored.
I can imagine that Core could keep a checkpoint counter and the Context would record last finished counter. During registration, if the current counter equals finished counter the registration could throw. However if we're to rollback the checkpoint, we would probably need to set a global flag in Core, too.

Reply:
Yes, the current implementation allows new Resources registered during notification and silently ignored. I hope this change will deal with the problem somehow.

> The best you could do is to invoke the beforeCheckpoint on the resource when you find C1 to be done, but this is not what we discussed before.

This is an interesting approach, but this won't suit e.g. Global Context, as last registered Resource would need to be notified first. If notification is in progress, we've lost our chance.

But I assume some Context implementation may opt to this, so we need to leave some room in the spec to allow such implementations.